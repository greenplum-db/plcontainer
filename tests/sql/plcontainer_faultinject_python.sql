-- Install a helper function to inject faults, using the fault injection
-- mechanism built into the server.
CREATE EXTENSION gp_inject_fault;

CREATE OR REPLACE FUNCTION pyint(i int) RETURNS int AS $$
# container: plc_python_shared
return i+1
$$ LANGUAGE plcontainer;

CREATE TABLE tbl(i int);

INSERT INTO tbl SELECT * FROM generate_series(1, 10);

-- start_ignore
-- QE crash after start a container 
SELECT gp_inject_fault('plcontainer_before_container_started', 'fatal', 2);
SELECT pyint(i) from tbl;
SELECT pg_sleep(5);
-- end_ignore

\! ssh `psql -d ${PL_TESTDB} -c 'select address from gp_segment_configuration where dbid=2' -t -A` docker ps -a | wc -l
\! ssh `psql -d ${PL_TESTDB} -c 'select address from gp_segment_configuration where dbid=2' -t -A` ps -ef | grep -v grep | grep "plcontainer cleaner" | wc -l
SELECT sum(pyint(i)) from tbl;

-- start_ignore
-- Start a container

-- QE crash when connnecting to an existing container
SELECT gp_inject_fault('plcontainer_before_container_connected', 'fatal', 2);
SELECT pyint(i) from tbl;
SELECT pg_sleep(5);
-- end_ignore

\! ssh `psql -d ${PL_TESTDB} -c 'select address from gp_segment_configuration where dbid=2' -t -A` docker ps -a | wc -l
\! ssh `psql -d ${PL_TESTDB} -c 'select address from gp_segment_configuration where dbid=2' -t -A` ps -ef | grep -v grep | grep "plcontainer cleaner" | wc -l
SELECT sum(pyint(i)) from tbl;

-- start_ignore
SELECT gp_inject_fault('plcontainer_after_send_request', 'fatal', 2);
SELECT pyint(i) from tbl;
SELECT pg_sleep(5);
-- end_ignore

\! ssh `psql -d ${PL_TESTDB} -c 'select address from gp_segment_configuration where dbid=2' -t -A` docker ps -a | wc -l
\! ssh `psql -d ${PL_TESTDB} -c 'select address from gp_segment_configuration where dbid=2' -t -A` ps -ef | grep -v grep | grep "plcontainer cleaner" | wc -l
SELECT sum(pyint(i)) from tbl;

-- start_ignore
SELECT gp_inject_fault('plcontainer_after_recv_request', 'fatal', 2);
SELECT pyint(i) from tbl;
SELECT pg_sleep(5);
-- end_ignore

\! ssh `psql -d ${PL_TESTDB} -c 'select address from gp_segment_configuration where dbid=2' -t -A` docker ps -a | wc -l
\! ssh `psql -d ${PL_TESTDB} -c 'select address from gp_segment_configuration where dbid=2' -t -A` ps -ef | grep -v grep | grep "plcontainer cleaner" | wc -l
SELECT sum(pyint(i)) from tbl;

-- start_ignore
SELECT gp_inject_fault('plcontainer_before_udf_finish', 'fatal', 2);
SELECT pyint(i) from tbl;
SELECT pg_sleep(5);
-- end_ignore

\! ssh `psql -d ${PL_TESTDB} -c 'select address from gp_segment_configuration where dbid=2' -t -A` docker ps -a | wc -l
\! ssh `psql -d ${PL_TESTDB} -c 'select address from gp_segment_configuration where dbid=2' -t -A` ps -ef | grep -v grep | grep "plcontainer cleaner" | wc -l

-- reset the injection points
SELECT gp_inject_fault('plcontainer_before_container_started', 'reset', 2);
SELECT gp_inject_fault('plcontainer_before_container_connected', 'reset', 2);
SELECT gp_inject_fault('plcontainer_after_send_request', 'reset', 2);
SELECT gp_inject_fault('plcontainer_after_recv_request', 'reset', 2);
SELECT gp_inject_fault('plcontainer_before_udf_finish', 'reset', 2);

-- After QE log(error, ...), related docker containers should be deleted.
-- start_ignore
SELECT gp_inject_fault('plcontainer_before_container_started', 'error', 1);
SELECT pyint(0);
SELECT pg_sleep(5);
-- end_ignore

\! docker ps -a | wc -l
\! ps -ef | grep -v grep | grep "plcontainer cleaner" | wc -l
SELECT pyint(0);

-- start_ignore
SELECT gp_inject_fault('plcontainer_after_send_request', 'error', 1);
SELECT pyint(0);
SELECT pg_sleep(5);
-- end_ignore

\! docker ps -a | wc -l
\! ps -ef | grep -v grep | grep "plcontainer cleaner" | wc -l
SELECT pyint(0);
-- Detect for the process name change (from "plcontainer cleaner" to other).
-- In such case, above cases will still succeed as unexpected.
\! docker ps -a | wc -l
\! ps -ef | grep -v grep | grep "plcontainer cleaner" | wc -l

-- reset the injection points
SELECT gp_inject_fault('plcontainer_before_container_started', 'reset', 1);
SELECT gp_inject_fault('plcontainer_after_send_request', 'reset', 1);
