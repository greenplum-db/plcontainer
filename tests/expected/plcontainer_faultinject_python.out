-- Install a helper function to inject faults, using the fault injection
-- mechanism built into the server.
CREATE EXTENSION gp_inject_fault;
CREATE OR REPLACE FUNCTION pyint(i int) RETURNS int AS $$
# container: plc_python_shared
return i+1
$$ LANGUAGE plcontainer;
CREATE TABLE tbl(i int);
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'i' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
INSERT INTO tbl SELECT * FROM generate_series(1, 10);
-- start_ignore
-- QE crash after start a container 
SELECT gp_inject_fault('plcontainer_before_container_started', 'fatal', 2);
NOTICE:  Success:
 gp_inject_fault 
-----------------
 t
(1 row)

SELECT pyint(i) from tbl;
ERROR:  fault triggered, fault name:'plcontainer_before_container_started' fault type:'fatal'  (seg0 slice1 127.0.0.1:25432 pid=15262)
SELECT pg_sleep(5);
 pg_sleep 
----------
 
(1 row)

-- end_ignore
\! docker ps -a | wc -l
1
\! ps -ef | grep "plcontainer cleaner" | grep -v pg_regress | wc -l
1
SELECT pyint(i) from tbl order by i;
 pyint 
-------
     2
     3
     4
     5
     6
     7
     8
     9
    10
    11
(10 rows)

-- start_ignore
-- Start a container
-- QE crash when connnecting to an existing container
SELECT gp_inject_fault('plcontainer_before_container_connected', 'fatal', 2);
NOTICE:  Success:
 gp_inject_fault 
-----------------
 t
(1 row)

SELECT pyint(i) from tbl;
ERROR:  fault triggered, fault name:'plcontainer_before_container_connected' fault type:'fatal'  (seg0 slice1 127.0.0.1:25432 pid=15363)
SELECT pg_sleep(5);
 pg_sleep 
----------
 
(1 row)

-- end_ignore
\! docker ps -a | wc -l
1
\! ps -ef | grep "plcontainer cleaner" | grep -v pg_regress | wc -l
1
SELECT pyint(i) from tbl order by i;
 pyint 
-------
     2
     3
     4
     5
     6
     7
     8
     9
    10
    11
(10 rows)

-- start_ignore
SELECT gp_inject_fault('plcontainer_after_send_request', 'fatal', 2);
NOTICE:  Success:
 gp_inject_fault 
-----------------
 t
(1 row)

SELECT pyint(i) from tbl;
ERROR:  fault triggered, fault name:'plcontainer_after_send_request' fault type:'fatal'  (seg0 slice1 127.0.0.1:25432 pid=15442)
SELECT pg_sleep(5);
 pg_sleep 
----------
 
(1 row)

-- end_ignore
\! docker ps -a | wc -l
1
\! ps -ef | grep "plcontainer cleaner" | grep -v pg_regress | wc -l
1
SELECT pyint(i) from tbl order by i;
 pyint 
-------
     2
     3
     4
     5
     6
     7
     8
     9
    10
    11
(10 rows)

-- start_ignore
SELECT gp_inject_fault('plcontainer_after_recv_request', 'fatal', 2);
NOTICE:  Success:
 gp_inject_fault 
-----------------
 t
(1 row)

SELECT pyint(i) from tbl;
ERROR:  fault triggered, fault name:'plcontainer_after_recv_request' fault type:'fatal'  (seg0 slice1 127.0.0.1:25432 pid=15522)
SELECT pg_sleep(5);
 pg_sleep 
----------
 
(1 row)

-- end_ignore
\! docker ps -a | wc -l
1
\! ps -ef | grep "plcontainer cleaner" | grep -v pg_regress | wc -l
1
SELECT pyint(i) from tbl order by i;
 pyint 
-------
     2
     3
     4
     5
     6
     7
     8
     9
    10
    11
(10 rows)

-- start_ignore
SELECT gp_inject_fault('plcontainer_before_udf_finish', 'fatal', 2);
NOTICE:  Success:
 gp_inject_fault 
-----------------
 t
(1 row)

SELECT pyint(i) from tbl;
ERROR:  fault triggered, fault name:'plcontainer_before_udf_finish' fault type:'fatal'  (seg0 slice1 127.0.0.1:25432 pid=15602)
SELECT pg_sleep(5);
 pg_sleep 
----------
 
(1 row)

-- end_ignore
\! docker ps -a | wc -l
1
\! ps -ef | grep "plcontainer cleaner" | grep -v pg_regress | wc -l
1
-- reset the injection points
SELECT gp_inject_fault('plcontainer_before_container_started', 'reset', 2);
NOTICE:  Success:
 gp_inject_fault 
-----------------
 t
(1 row)

SELECT gp_inject_fault('plcontainer_before_container_connected', 'reset', 2);
NOTICE:  Success:
 gp_inject_fault 
-----------------
 t
(1 row)

SELECT gp_inject_fault('plcontainer_after_send_request', 'reset', 2);
NOTICE:  Success:
 gp_inject_fault 
-----------------
 t
(1 row)

SELECT gp_inject_fault('plcontainer_after_recv_request', 'reset', 2);
NOTICE:  Success:
 gp_inject_fault 
-----------------
 t
(1 row)

SELECT gp_inject_fault('plcontainer_before_udf_finish', 'reset', 2);
NOTICE:  Success:
 gp_inject_fault 
-----------------
 t
(1 row)

-- After QE log(error, ...), related docker containers should be deleted.
-- start_ignore
SELECT gp_inject_fault('plcontainer_before_container_started', 'error', 1);
NOTICE:  Success:
 gp_inject_fault 
-----------------
 t
(1 row)

SELECT pyint(0);
ERROR:  fault triggered, fault name:'plcontainer_before_container_started' fault type:'error'
SELECT pg_sleep(5);
 pg_sleep 
----------
 
(1 row)

-- end_ignore
\! docker ps -a | wc -l
1
\! ps -ef | grep "plcontainer cleaner" | grep -v pg_regress | wc -l
1
SELECT pyint(0);
 pyint 
-------
     1
(1 row)

-- start_ignore
SELECT gp_inject_fault('plcontainer_after_send_request', 'error', 1);
NOTICE:  Success:
 gp_inject_fault 
-----------------
 t
(1 row)

SELECT pyint(0);
ERROR:  fault triggered, fault name:'plcontainer_after_send_request' fault type:'error'
SELECT pg_sleep(5);
 pg_sleep 
----------
 
(1 row)

-- end_ignore
\! docker ps -a | wc -l
1
\! ps -ef | grep "plcontainer cleaner" | grep -v pg_regress | wc -l
1
SELECT pyint(0);
 pyint 
-------
     1
(1 row)

-- reset the injection points
SELECT gp_inject_fault('plcontainer_before_container_started', 'reset', 1);
NOTICE:  Success:
 gp_inject_fault 
-----------------
 t
(1 row)

SELECT gp_inject_fault('plcontainer_after_send_request', 'reset', 1);
NOTICE:  Success:
 gp_inject_fault 
-----------------
 t
(1 row)

